service-base/src/main/java/org/eclipse/hono/service/plan/PrometheusBasedResourceLimitChecks.java

```diff
@@ -261,19 +339,74 @@ private Long extractLongValue(final JsonObject response) {
     }
 
     /**
-     * Calculate the period for which the data usage is to be retrieved from the prometheus server.
+     * Calculates the data volume (in bytes) that is allowed for a tenant for 
+     * the current period based on the mode defined by 
+     * {@link TenantConstants#FIELD_PERIOD_MODE}.
+     * <p>
+     * In <em>monthly</em> mode, if the effectiveSince date doesn't fall on the 
+     * first day of the month then the allowed data volume for the tenant is 
+     * calculated as below. For rest of the months and the <em>days</em> mode,
+     * the maxBytes defined by {@link TenantConstants#FIELD_MAX_BYTES} is used directly.
+     * <pre>
+     *             maxBytes 
+     *   ---------------------------------- x No. of days from effectiveSince till lastDayOfCurrentMonth.
+     *    No. of days in the current month
+     * </pre>
+     * <p>
      * 
-     * @param effectiveSince The instant on which the data volume limit came into effect.
-     * @param periodInDays The number of days for which the data usage is to be calculated.
-     * @return The period for which the data usage is to be calculated.
+     * @param effectiveSince The point of time on which the data volume limit 
+     *                       came into effect.
+     * @param mode The mode of the period. 
+     * @param maxBytes The maximum allowed bytes defined in configuration by 
+     *                 {@link TenantConstants#FIELD_MAX_BYTES}. 
+     * @return The allowed data-volume in bytes.
      */
-    private long calculateDataUsagePeriod(final Instant effectiveSince, final long periodInDays) {
-        final long inclusiveDaysBetween = DAYS.between(effectiveSince, Instant.now()) + 1;
-        if (inclusiveDaysBetween > 0 && periodInDays > 0) {
-            final long dataUsagePeriodInDays = inclusiveDaysBetween % periodInDays;
-            return dataUsagePeriodInDays == 0 ? periodInDays : dataUsagePeriodInDays;
+    private long calculateDataVolume(final OffsetDateTime effectiveSince, final PeriodMode mode,
```

Comment:
I think we should accept an `Instant` instead of an `OffsetDateTime` here ..

Reply:
Since we don't serialize/de-serialize here, I thought its fine to use the `OffsetDateTime`. `OffsetDateTime` comes with functions like `getDayOfMonth` which is quite handy to perform calculations.

Reply:
`Instant.get(ChronoField.DAY_OF_MONTH)`

Reply:
`Instant.now().get(ChronoField.DAY_OF_MONTH)` throws
`Exception in thread "main" java.time.temporal.UnsupportedTemporalTypeException: Unsupported field: DayOfMonth`. It also the same with YearMonth.from(Instant.now()).

It seems a zone or offset is needed for it as below:
`Instant.now().atOffset(ZoneOffset.UTC).get(ChronoField.DAY_OF_MONTH)`

Reply:
Instead of converting `Instant` to offset during every such cases as above, I would prefer leaving it as `OffsetDateTime` instead of `Instant`. WDYT?

Reply:
IMHO, using `OffsetDateTime` looks ok here. Quoting the `OffsetDateTime` javadoc:
`OffsetDateTime adds to the instant the offset from UTC/Greenwich, which allows the local date-time to be obtained.` And by means of the local date, the day of month can be obtained (via `OffsetDateTime.get(ChronoField.DAY_OF_MONTH)` or `OffsetDateTime.getDayOfMonth()`).

Reply:
> IMHO, using `OffsetDateTime` looks ok here. Quoting the `OffsetDateTime` javadoc:
> `OffsetDateTime adds to the instant the offset from UTC/Greenwich, which allows the local date-time to be obtained.` And by means of the local date, the day of month can be obtained (via `OffsetDateTime.get(ChronoField.DAY_OF_MONTH)` or `OffsetDateTime.getDayOfMonth()`).

:+1: I have pushed changes corresponding to the other review comments and also changes to store the calculated max bytes and data volume period in a cache instead of calculating every time.